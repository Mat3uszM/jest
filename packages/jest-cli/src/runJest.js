/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @flow
 */
'use strict';

import type {Context} from 'types/Context';
import type {Test} from 'types/TestRunner';
import type {PatternInfo} from './SearchSource';
import type TestWatcher from './TestWatcher';

const fs = require('graceful-fs');

const SearchSource = require('./SearchSource');
const TestRunner = require('./TestRunner');
const TestSequencer = require('./TestSequencer');

const getTestPathPatternInfo = require('./lib/getTestPathPatternInfo');
const chalk = require('chalk');
const {Console, formatTestResults} = require('jest-util');
const getMaxWorkers = require('./lib/getMaxWorkers');
const path = require('path');
const setState = require('./lib/setState');

const getTestSummary = (argv: Object, patternInfo: PatternInfo) => {
  const testPathPattern = SearchSource.getTestPathPattern(patternInfo);
  const testInfo = patternInfo.onlyChanged
    ? chalk.dim(' related to changed files')
    : patternInfo.input !== '' ? chalk.dim(' matching ') + testPathPattern : '';

  const nameInfo = argv.testNamePattern
    ? chalk.dim(' with tests matching ') + `"${argv.testNamePattern}"`
    : '';

  return chalk.dim('Ran all test suites') +
    testInfo +
    nameInfo +
    chalk.dim('.');
};

const getTestPaths = async (context, patternInfo, argv, pipe) => {
  const source = new SearchSource(context, context.config);
  let data = await source.getTestPaths(patternInfo);
  if (!data.paths.length) {
    const localConsole = new Console(pipe, pipe);
    if (patternInfo.onlyChanged && data.noSCM) {
      if (context.config.watch) {
        // Run all the tests
        setState(argv, 'watchAll', {
          noSCM: true,
        });
        patternInfo = getTestPathPatternInfo(argv);
        data = await source.getTestPaths(patternInfo);
      } else {
        localConsole.log(
          'Jest can only find uncommitted changed files in a git or hg ' +
            'repository. If you make your project a git or hg ' +
            'repository (`git init` or `hg init`), Jest will be able ' +
            'to only run tests related to files changed since the last ' +
            'commit.',
        );
      }
    }

    localConsole.log(
      source.getNoTestsFoundMessage(patternInfo, context.config, data),
    );
  }

  return {
    data,
    patternInfo,
  };
};

const runJest = async (
  contexts: Array<Context>,
  argv: Object,
  pipe: stream$Writable | tty$WriteStream,
  testWatcher: TestWatcher,
  startRun: () => *,
  onComplete: (testResults: any) => void,
) => {
  const maxWorkers = getMaxWorkers(argv);
  const context = contexts[0];
  const testRunData = await Promise.all(
    contexts.map(async context => {
      const {data, patternInfo} = await getTestPaths(
        context,
        getTestPathPatternInfo(argv),
        argv,
        pipe,
      );

      const sequencer = new TestSequencer(context);
      const tests = sequencer.sort(data.paths);
      return {context, patternInfo, sequencer, tests};
    }),
  );

  const processResults = runResults => {
    if (context.config.testResultsProcessor) {
      /* $FlowFixMe */
      runResults = require(context.config.testResultsProcessor)(runResults);
    }
    if (argv.json) {
      if (argv.outputFile) {
        const outputFile = path.resolve(process.cwd(), argv.outputFile);

        fs.writeFileSync(
          outputFile,
          JSON.stringify(formatTestResults(runResults)),
        );
        process.stdout.write(
          `Test results written to: ` +
            `${path.relative(process.cwd(), outputFile)}\n`,
        );
      } else {
        process.stdout.write(JSON.stringify(formatTestResults(runResults)));
      }
    }
    return onComplete && onComplete(runResults);
  };

  const allTests = testRunData
    .reduce((tests, testRun) => tests.concat(testRun.tests), [])
    .sort((a: Test, b: Test) => {
      if (a.duration != null && b.duration != null) {
        return a.duration < b.duration ? 1 : -1;
      }
      return a.duration == null ? 1 : 0;
    });

  if (
    allTests.length === 1 &&
    context.config.silent !== true &&
    context.config.verbose !== false
  ) {
    context.config = Object.assign({}, context.config, {verbose: true});
  }

  if (context.config.updateSnapshot === true) {
     context.config = Object.assign({}, context.config, {
      updateSnapshot: true,
    });
  }

  const results = await new TestRunner(
    context,
    context.config,
    {
      getTestSummary: () => getTestSummary(argv, testRunData[0].patternInfo),
      maxWorkers,
    },
    startRun,
  ).runTests(allTests, testWatcher);
  testRunData.forEach(({sequencer, tests}) =>
    sequencer.cacheResults(tests, results));
  return processResults(results);
};

module.exports = runJest;
