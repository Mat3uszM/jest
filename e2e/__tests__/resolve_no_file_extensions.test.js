/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */
'use strict';

import runJest from '../runJest';
import {extractSummary} from '../Utils';

test('show error message with matching files', () => {
  const {status, stderr} = runJest('resolve_no_extensions');
  const {rest} = extractSummary(stderr);

  expect(status).toBe(1);
  expect(rest).toMatchSnapshot();
});
