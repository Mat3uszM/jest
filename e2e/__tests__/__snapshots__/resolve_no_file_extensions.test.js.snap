// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`show error message when no js moduleFileExtensions 1`] = `
"● Validation Error:

  moduleFileExtensions must include 'js':
  but instead received:
    [\\"jsx\\"]
  Please change your configuration to include 'js'.

  Configuration Documentation:
  https://jestjs.io/docs/configuration.html

"
`;

exports[`show error message with matching files 1`] = `
"FAIL __tests__/test.js
  ● Test suite failed to run

    Cannot find module './some-json-file' from 'index.js'

    However, Jest was able to find:
    	'./some-json-file.json'

    You might want to include a file extension in your import, or update your 'moduleFileExtensions', which is currently ['js'].

    See https://jestjs.io/docs/en/configuration#modulefileextensions-array-string

    > 1 | module.exports = require('./some-json-file');
        |                  ^
      2 | 

      at packages/jest-resolve/build/index.js:221:17
      at index.js:1:18

"
`;
